designer/server/src/main/scala/pl/touk/nussknacker/ui/process/processingtype/provider/ProcessingTypeDataProvider.scala

```diff
@@ -86,112 +88,85 @@ trait ProcessingTypeDataProvider[+Data, +CombinedData] {
 
   // TODO: We should return a generic type that can produce views for users with access rights to certain categories only.
   //       Thanks to that we will be sure that no sensitive data leak
-  final def combined: CombinedData = state.getCombined()
-
-  private[processingtype] def state: ProcessingTypeDataState[Data, CombinedData]
-
-  final def mapValues[TT](fun: Data => TT): ProcessingTypeDataProvider[TT, CombinedData] =
-    new TransformingProcessingTypeDataProvider[Data, CombinedData, TT, CombinedData](this, _.mapValues(fun))
-
-  final def mapCombined[CC](fun: CombinedData => CC): ProcessingTypeDataProvider[Data, CC] =
-    new TransformingProcessingTypeDataProvider[Data, CombinedData, Data, CC](this, _.mapCombined(fun))
-
-}
-
-private[processingtype] class TransformingProcessingTypeDataProvider[T, C, TT, CC](
-    observed: ProcessingTypeDataProvider[T, C],
-    transformState: ProcessingTypeDataState[T, C] => ProcessingTypeDataState[TT, CC]
-) extends ProcessingTypeDataProvider[TT, CC] {
-
-  private val stateValue = new AtomicReference(transformState(observed.state))
+  final def combined: CombinedData = state.combinedDataTry.get
+
+  final protected[provider] def setStateValueAndNotifyObservers(
+      newValue: ProcessingTypeDataState[Data, CombinedData]
+  ): IO[Unit] = {
+    for {
+      _ <- stateRef.set(newValue)
+      _ <- observers.asScala
+        .map { observer =>
+          observer.notifyChange(newValue)
+        }
+        .toList
+        .sequence
+    } yield ()
+  }
 
-  override private[processingtype] def state: ProcessingTypeDataState[TT, CC] = {
-    stateValue.updateAndGet { currentValue =>
-      val currentObservedState = observed.state
-      if (currentObservedState.stateIdentity != currentValue.stateIdentity) {
-        transformState(currentObservedState)
-      } else {
-        currentValue
+  private def state: ProcessingTypeDataState[Data, CombinedData] = {
+    stateMutex.lock
+      .surround {
+        stateRef.get
       }
-    }
+      .unsafeRunSync()(IORuntime.global) // TODO: get rid of unsafeRunSync
```

Comment:
This is tricky after this changes with mutex and ref.  This method is used very often, during almost every request processing.

Reply:
ok, so it's related to this topic:
https://github.com/TouK/nussknacker/pull/7588/files#r1986981795

The problem is that the child class's implementation `reloadAll` method affects the internals of the base class.

Reply:
oh, and when we do `unsafeRunSync` (I agree that refactor the app to not have to do that is not so easy right now), it'd be better to do it on a separate IORuntime (not the global one).